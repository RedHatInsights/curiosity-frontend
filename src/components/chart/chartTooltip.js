import React from 'react';
import { helpers } from '../../common';

/**
 * Note: Victory Charts components require a form of extended "something" applied by the consuming component.
 * This leads to odd implementations, like returning a function component.
 */
/**
 * Return a compatible Victory tooltip component.
 *
 * @param {object} params
 * @param {object} params.chartSettings
 * @param {Function} params.chartContainerRef
 * @param {Function} params.chartTooltipRef
 * @param {number} params.minChartWidth
 * @returns {Function}
 */
const chartTooltip = ({
  chartSettings = {},
  chartContainerRef = helpers.noop,
  chartTooltipRef = helpers.noop,
  minChartWidth = 500
} = {}) => {
  // ToDo: evaluate using "width" in place of "minWidth" for scenarios where the graph size is smaller
  /**
   * Return a tooltip x coordinate.
   *
   * @param {object} params
   * @param {number} params.x
   * @param {number} params.width
   * @param {number} params.tooltipWidth
   * @param {number} params.padding
   * @param {number} params.minWidth
   * @returns {number}
   */
  const getXCoordinate = ({ x, width, tooltipWidth, padding = 0, minWidth = minChartWidth } = {}) => {
    if (width <= minWidth && x > tooltipWidth / 2 + padding && x < minWidth - tooltipWidth + padding) {
      return x + padding - tooltipWidth / 2;
    }

    return x > width / 2 ? x - tooltipWidth + padding : x + padding;
  };

  /**
   * Return a tooltip y coordinate.
   *
   * @param {object} params
   * @param {number} params.y
   * @param {number} params.height
   * @param {number} params.tooltipHeight
   * @param {number} params.width
   * @param {number} params.padding
   * @param {number} params.minWidth
   * @returns {number}
   */
  const getYCoordinate = ({ y, height, tooltipHeight, width, padding = 15, minWidth = minChartWidth } = {}) => {
    if (width <= minWidth) {
      return y > height / 2 ? y - tooltipHeight - padding : y + padding;
    }

    return height * 0.25;
  };

  // ToDo: evaluate using "width" in place of "minWidth" for scenarios where the graph size is smaller
  /**
   * Return a tooltip tail position CSS class.
   *
   * @param {object} params
   * @param {number} params.x
   * @param {number} params.width
   * @param {number} params.tooltipWidth
   * @param {number} params.padding
   * @param {number} params.minWidth
   * @returns {number}
   */
  const tailPosition = ({ x, width, tooltipWidth, padding = 0, minWidth = minChartWidth } = {}) => {
    if (width <= minWidth && x > tooltipWidth / 2 + padding && x < minWidth - tooltipWidth + padding) {
      return 'middle';
    }

    return x > width / 2 ? 'right' : 'left';
  };

  return ({ x, y, datum = {} }) => { // eslint-disable-line
    const { padding = {}, tooltipDataSetLookUp = {} } = chartSettings;

    const containerRef = chartContainerRef();
    const tooltipRef = chartTooltipRef();
    const content = tooltipDataSetLookUp?.[datum.x]?.tooltip || '';
    const containerPaddingBottom = padding.bottom ?? 0;
    const containerBounds = containerRef?.current?.querySelector('svg')?.getBoundingClientRect() || {
      width: 0,
      height: 0
    };
    const tooltipBounds = tooltipRef?.current?.getBoundingClientRect() || { width: 0, height: 0 };

    if (content) {
      const updatedClassName = `${(tooltipBounds.height <= 0 && 'fadein') || ''}`;

      return (
        <g>
          <foreignObject
            x={getXCoordinate({ x, width: containerBounds.width, tooltipWidth: tooltipBounds.width })}
            y={getYCoordinate({
              y,
              height: containerBounds.height,
              tooltipHeight: tooltipBounds.height,
              width: containerBounds.width
            })}
            width="100%"
            height="100%"
          >
            <div
              className={`curiosity-chartarea__tooltip-container ${updatedClassName}`}
              ref={tooltipRef}
              style={{ display: (y > containerBounds.height - containerPaddingBottom && 'none') || 'inline-block' }}
              xmlns="http://www.w3.org/1999/xhtml"
            >
              <div
                className={`curiosity-chartarea__tooltip curiosity-chartarea__tooltip-${tailPosition({
                  x,
                  y,
                  width: containerBounds.width,
                  tooltipWidth: tooltipBounds.width
                })}`}
              >
                {content}
              </div>
            </div>
          </foreignObject>
        </g>
      );
    }

    return <g />;
  };
};

export { chartTooltip as default, chartTooltip };
