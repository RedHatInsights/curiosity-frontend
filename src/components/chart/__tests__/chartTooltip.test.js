import React from 'react';
import { chartTooltip } from '../chartTooltip';

describe('ChartTooltip Component', () => {
  it('should return a basic component', () => {
    const settings = {
      chartSettings: {
        padding: {},
        tooltipDataSetLookUp: {
          1: {
            tooltip: 'lorem ipsum 1'
          },
          2: {
            tooltip: 'dolor sit 2'
          }
        }
      }
    };

    const props = {
      x: 2,
      y: 3,
      datum: { x: 1 }
    };

    const Component = chartTooltip(settings);
    const updatedComponent = renderComponent(<Component {...props} />);

    expect(updatedComponent).toMatchSnapshot('basic');
  });

  it('should return an empty tooltip when missing content', () => {
    const settings = {
      chartSettings: {
        padding: {},
        tooltipDataSetLookUp: {
          1: {
            tooltip: undefined
          },
          2: {
            tooltip: undefined
          }
        }
      }
    };

    const props = {
      x: 2,
      y: 3,
      datum: { x: 1 }
    };

    const Component = chartTooltip(settings);
    const updatedComponent = renderComponent(<Component {...props} />);

    expect(updatedComponent).toMatchSnapshot('missing content');
  });

  it('should handle tooltip position', () => {
    const settings = {
      chartContainerRef: () => ({
        current: { querySelector: () => ({ getBoundingClientRect: () => ({ width: 500, height: 100 }) }) }
      }),
      chartTooltipRef: () => ({ current: { getBoundingClientRect: () => ({ width: 40, height: 15 }) } }),
      chartSettings: {
        padding: {},
        tooltipDataSetLookUp: {
          1: {
            tooltip: 'lorem ipsum 1'
          },
          2: {
            tooltip: 'dolor sit 2'
          },
          3: {
            tooltip: 'hello world 3'
          }
        }
      }
    };

    const props = {
      x: 0,
      y: 75,
      datum: { x: 1 }
    };

    const Component = chartTooltip(settings);
    const updatedComponent = renderComponent(<Component {...props} />);

    expect(updatedComponent).toMatchSnapshot('left tail');

    const middleTail = updatedComponent.setProps({
      x: 250,
      datum: { x: 2 }
    });

    expect(middleTail).toMatchSnapshot('middle tail');

    const rightTail = updatedComponent.setProps({
      x: 500,
      datum: { x: 3 }
    });

    expect(rightTail).toMatchSnapshot('right tail');
  });
});
