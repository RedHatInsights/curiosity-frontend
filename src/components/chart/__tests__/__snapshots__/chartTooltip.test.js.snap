// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTooltip Component should handle tooltip position: left tail 1`] = `
<element
  props="{
  "x": 0,
  "y": 75,
  "datum": {
    "x": 1
  },
  "children": []
}"
>
  <g>
    <foreignobject
      height="100%"
      width="100%"
      x="0"
      y="45"
    >
      <div
        class="curiosity-chartarea__tooltip-container "
        style="display: inline-block;"
        xmlns="http://www.w3.org/1999/xhtml"
      >
        <div
          class="curiosity-chartarea__tooltip curiosity-chartarea__tooltip-left"
        >
          lorem ipsum 1
        </div>
      </div>
    </foreignobject>
  </g>
</element>
`;

exports[`ChartTooltip Component should handle tooltip position: middle tail 1`] = `
<element
  props="{
  "x": 250,
  "y": 75,
  "datum": {
    "x": 2
  },
  "children": []
}"
>
  <g>
    <foreignobject
      height="100%"
      width="100%"
      x="230"
      y="45"
    >
      <div
        class="curiosity-chartarea__tooltip-container "
        style="display: inline-block;"
        xmlns="http://www.w3.org/1999/xhtml"
      >
        <div
          class="curiosity-chartarea__tooltip curiosity-chartarea__tooltip-middle"
        >
          dolor sit 2
        </div>
      </div>
    </foreignobject>
  </g>
</element>
`;

exports[`ChartTooltip Component should handle tooltip position: right tail 1`] = `
<element
  props="{
  "x": 500,
  "y": 75,
  "datum": {
    "x": 3
  },
  "children": []
}"
>
  <g>
    <foreignobject
      height="100%"
      width="100%"
      x="460"
      y="45"
    >
      <div
        class="curiosity-chartarea__tooltip-container "
        style="display: inline-block;"
        xmlns="http://www.w3.org/1999/xhtml"
      >
        <div
          class="curiosity-chartarea__tooltip curiosity-chartarea__tooltip-right"
        >
          hello world 3
        </div>
      </div>
    </foreignobject>
  </g>
</element>
`;

exports[`ChartTooltip Component should return a basic component: basic 1`] = `
<element
  props="{
  "x": 2,
  "y": 3,
  "datum": {
    "x": 1
  },
  "children": []
}"
>
  <g>
    <foreignobject
      height="100%"
      width="100%"
      x="2"
      y="-12"
    >
      <div
        class="curiosity-chartarea__tooltip-container fadein"
        style="display: none;"
        xmlns="http://www.w3.org/1999/xhtml"
      >
        <div
          class="curiosity-chartarea__tooltip curiosity-chartarea__tooltip-middle"
        >
          lorem ipsum 1
        </div>
      </div>
    </foreignobject>
  </g>
</element>
`;

exports[`ChartTooltip Component should return an empty tooltip when missing content: missing content 1`] = `
<element
  props="{
  "x": 2,
  "y": 3,
  "datum": {
    "x": 1
  },
  "children": []
}"
>
  <g />
</element>
`;
